scripts/components/git_client.ts

```diff
@@ -0,0 +1,212 @@
+  };
+
+  checkout = async (ref: string, paths: string[] = []) => {
+    const additionalArgs = paths.length > 0 ? ['--', ...paths] : [];
+    await this.execWithIO('git', ['checkout', ref, ...additionalArgs]);
+  };
+
+  status = async () => {
+    await this.execWithIO('git', ['status']);
+  };
+
+  /**
+   * Returns a list of tags that point to the given commit
+   */
+  getTagsAtCommit = async (commitHash: string) => {
+    // filter out packages not published to npm
+    const privatePackages = [
+      'amplify-category-api-dynamodb-simulator',
+      'amplify-category-api-e2e-core',
+      'amplify-category-api-e2e-tests',
+      'amplify-graphql-api-construct-tests',
+      'amplify-category-api-graphql-migration-tests',
+      '@aws-amplify/graphql-transformer-test-utils',
+      '@aws-amplify/graphql-schema-validation',
+      'amplify-category-api-util-mock',
+      'amplify-category-api-graphql-transformers-e2e-tests',
+    ];
+    const { stdout: tagsString } = await this.exec('git', ['tag', '--points-at', commitHash]);
+    return tagsString
+      .split(EOL)
+      .filter((line: string) => line.trim().length > 0)
+      .filter((tag: string) => !privatePackages.some((name) => tag.includes(name)));
+  };
+
+  /**
+   * Gets the most recent release commit that is reachable from the input commitHash
+   * If no commitHash is specified, HEAD is used as the default
+   * By default, the input commitHash is considered in the search (ie if commitHash is a release commit, that commit will be returned)
+   * To search for the most recent release commit EXCLUDING commitHash, set inclusive=false
+   */
+  getNearestReleaseCommit = async (commitHash: string = 'HEAD', { inclusive }: { inclusive: boolean } = { inclusive: true }) => {
+    // get the most recent tag before (or at if inclusive=false) the current release tag
+    const { stdout: previousReleaseTag } = await this.exec('git', ['describe', `${commitHash}${inclusive ? '' : '^'}`, '--abbrev=0']);
+
+    // get the commit hash associated with the previous release tag
+    const { stdout: previousReleaseCommitHash } = await this.exec('git', ['log', '-1', previousReleaseTag, '--pretty=%H']);
+
+    // run some sanity checks on the release commit
+    await this.validateReleaseCommitHash(previousReleaseCommitHash);
+
+    return previousReleaseCommitHash;
+  };
+
+  /**
+   * Given a release commit hash A that has tags for one or more package versions,
+   * walk through release history and find the previous release tags of all of the packages that were released in commit A
+   *
+   * Note that this does not mean just looking up the previous release tags.
+   * It may be the case that package-A was released in release-5 but the previous release of package-A happened in release-2.
+   * This method will walk through past release tags until it finds the previous version of all of the input package versions
+   * If a previous version of some package cannot be found, an error is thrown.
```

Comment:
DOes this mean that we are unable to roll back releases that contain the first public publication of a package?

Reply:
Yes, I'll see if there is a way to resolve this.

Reply:
I've changed this to just print a warning message when a previous release is not found.